src/test/client/TestNewOpsClient.h

```diff
@@ -0,0 +1,105 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2021 Red Hat
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "gtest/gtest.h"
+
+#include "common/async/context_pool.h"
+#include "global/global_context.h"
+
+#include "msg/Messenger.h"
+#include "mon/MonClient.h"
+#include "osdc/ObjectCacher.h"
+#include "client/MetaRequest.h"
+#include "client/Client.h"
+
+#define dout_subsys ceph_subsys_client
+
+class TestNewOpsClient : public Client {
+public:
+    //TestNewOpsClient();
+    TestNewOpsClient(Messenger *m, MonClient *mc, Objecter *objecter_) : Client(m, mc, objecter_) {}
+    ~TestNewOpsClient();
+    int check_dummy_op(const UserPerm& perms){
+      RWRef_t mref_reader(mount_state, CLIENT_MOUNTING);
+      if (!mref_reader.is_state_satisfied()) {
+        return -CEPHFS_ENOTCONN;
+      }
+      MetaRequest *req = new MetaRequest(CEPH_MDS_OP_DUMMY);
+      int res = make_request(req, perms);
+      ldout(cct, 10) << __func__ << " result=" << res << dendl;
+      return res;
+    }
+};
+
+class TestNewOps : public ::testing::Test { 
```

Comment:
This class is copied from `TestClient`. How about adding a template for this ?

```
template<typename T>
class TestNewOps : public ::testing::Test {
...
client = new T(messenger, mc, objecter);
...
T* client = nullptr;
}
```

Reply:
I tried this up and it I hit issue that says `object of abstract class type "TestClient<TestNewOpsClient>" is not allowed:C/C++(322)
ops.cc(21, 1): pure virtual function "testing::Test::TestBody" has no overrider`

I tried looking at `gtest.h` which is being used in this file, it has virtual functions, so does that mean I cannot use template in a class that derives a class containing virtual functions? Any idea anyone?

Reply:
So the `gtest.h` has a pure virtual function `TestBody()`, which it seems is not compatible with template creation. I'm not cent percent sure about, can anyone opine/suggest something that would work?

Reply:
Closing this conversation as it seems that using template when base class has pure virtual function(s) is not allowed